Source/AwsCommonRuntimeKit/event-stream/EventStreamMessage.swift

```diff
@@ -0,0 +1,137 @@
+        }
+        try headers.forEach {
+            try addHeader(header: $0, rawHeaders: &rawHeaders)
+        }
+
+        guard payload.withAWSByteBufPointer({ byteBuff in
+            // TODO (optimization): we could avoid the extra copies of headers and data
+            // if there were an API in C that let us encode everything directly into a pre-allocated buffer
+            aws_event_stream_message_init(&rawValue, allocator.rawValue, &rawHeaders, byteBuff)
+        }) == AWS_OP_SUCCESS else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+
+        return Data(
+            bytes: aws_event_stream_message_buffer(&rawValue),
+            count: Int(aws_event_stream_message_total_length(&rawValue)))
+    }
+}
+
+extension EventStreamMessage {
+    func addHeader(header: EventStreamHeader, rawHeaders: UnsafeMutablePointer<aws_array_list>) throws {
+        if header.name.count > EventStreamHeader.maxNameLength {
+            throw CommonRunTimeError.crtError(
+                .init(
+                    code: AWS_ERROR_EVENT_STREAM_MESSAGE_INVALID_HEADERS_LEN.rawValue))
+        }
+        let addCHeader: () throws -> Int32 = {
+            return try header.name.withCString { headerName in
+                switch header.value {
+                case .bool(let value):
+                    return aws_event_stream_add_bool_header(
+                        rawHeaders,
+                        headerName,
+                        UInt8(header.name.count),
+                        Int8(value.uintValue))
+                case .byte(let value):
+                    return aws_event_stream_add_byte_header(
+                        rawHeaders,
+                        headerName,
+                        UInt8(header.name.count),
+                        value)
+                case .int16(let value):
+                    return aws_event_stream_add_int16_header(
+                        rawHeaders,
+                        headerName,
+                        UInt8(header.name.count),
+                        value)
+                case .int32(let value):
+                    return aws_event_stream_add_int32_header(
+                        rawHeaders,
+                        headerName,
+                        UInt8(header.name.count),
+                        value)
+                case .int64(let value):
+                    return aws_event_stream_add_int64_header(
+                        rawHeaders,
+                        headerName,
+                        UInt8(header.name.count),
+                        value)
+                case .byteBuf(var value):
+                    if value.count > EventStreamHeader.maxValueLength {
+                        throw CommonRunTimeError.crtError(
+                            .init(
+                                code: AWS_ERROR_EVENT_STREAM_MESSAGE_INVALID_HEADERS_LEN.rawValue))
+                    }
```

Comment:
should this check be in the C code?

So, consumer can throw the C emitted error and avoid duplicating this logic?

Reply:
There are already checks for this in C api. Problem is that it is defined in C as follow.

```
char *value;
int8 value_length;
```
So we need to cast `value.count` to `Int8` before passing to C, which is a crash in Swift if it is greater than Int8.max value. This check is there to make it a runtime error instead of crash.